File: src/main/java/core/procedures/strings/Substring.java
package core.procedures.strings;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.SCMClass;

public final class Substring extends AFn {

  public Substring() {
    super(new FnArgsBuilder().min(2).max(3)
                             .mandatory(new Class[]{CharSequence.class, SCMClass.ExactNonNegativeInteger.class})
                             .rest(SCMClass.ExactNonNegativeInteger.class).build());
  }

  @Override
  public String getName() {
    return "substring";
  }

  @Override
  public String apply(Object... args) {
    String s = args[0].toString();
    long start = ((Number)args[1]).longValue();
    if (start > s.length()) {
      throw new IndexOutOfBoundsException(String.format("%s: value out of range: %s", getName(), start));
    }

    long end = s.length();
    if (args.length == 3) {
      end = (long) args[2];
    }
    if (end > s.length()) {
      throw new IndexOutOfBoundsException(String.format("%s: value out of range: %s", getName(), end));
    }
    return s.substring((int)start, (int)end);
  }
}

File: src/main/java/core/procedures/strings/Substring.kt
package core.procedures.strings

import core.procedures.AFn
import core.scm.Type

open class Substring : AFn<Any?, String>(name = "substring", isPure = true, minArgs = 2, maxArgs = 3,
                           mandatoryArgsTypes = arrayOf(CharSequence::class.java, Type.ExactNonNegativeInteger::class.java),
                           restArgsType = Type.ExactNonNegativeInteger::class.java) {

    override operator fun invoke(args: Array<Any?>): String {
        val s = args[0].toString()
        val start = (args[1] as Number).toInt()
        if (start > s.length) {
            throw IndexOutOfBoundsException("$name: value out of range: $start")
        }
        var end = s.length
        if (args.size == 3) {
            end = (args[2] as Number).toInt()
        }
        if (end > s.length) {
            throw IndexOutOfBoundsException("$name: value out of range: $end")
        }
        return s.substring(start, end)
    }
}
